### PR TITLE
Feature/unr 4260 add worker info to boundary

### PR DESCRIPTION
**Pull request by Vix2021**
#### Description
The task was to add worker names and virtual worker ids to the worker boundaries. This was achieved by rendering a dynamic material that combines the background colour and the foreground worker information and repeat the material across the worker boundary faces.

Handover Gym:
DynamicMaterialFinal
Teleport Gym:
DynamicMaterialFinal2

Blueprint changes:
TranslucentWorkerRegion (minor change): Change to unlit material to avoid gradient which is visible when tiling the material
translucentworkerregionBP
WorkerRegionCombinedMaterial (new material):
Overview:
image
Part 1:
workerregioncombinedmaterialBP-p1
Part 2:
workerregioncombinedmaterialBP-p2
MuliFont (new Font)
mulifont
BP_SpatialDebugger (changes no longer needed, manually undid changes but do not know how to revert this asset to its previous state)

#### Release note
- You can now see worker information displayed on the worker's boundaries. The worker name and virtual worker id displayed corresponds to the worker that you are currently looking at and will be visible when you are near a border.

#### Tests
Tested in the GDK test gyms: handover and teleport with different load balancing strategies.
Tested in NWX to check visibility with the different day and night cycles.

STRONGLY SUGGESTED: How can this be verified by QA?
Needs corresponding GDK Test Gym PR to verify the changes: https://github.com/spatialos/UnrealGDKTestGyms/pull/212
Run a zoned map and check that the boundary information is visible and correct for each worker.

#### Documentation
Will update images of the worker boundaries in the multi worker document: https://docs.google.com/document/d/1G9cRz891Ycaiax-M7Psx3rlfX8qTIi0CTk8_71PL4Zg/edit#heading=h.j4ppr52lw6ck

#### Known issues
Pre-merge failing due to assets created in Unreal V.25. Re-creating them now in V.24.

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/WorkerRegion.cpp [hunk 1]
```diff
@@ -23,16 +26,56 @@ AWorkerRegion::AWorkerRegion(const FObjectInitializer& ObjectInitializer)
...
+	// Background translucent coloured worker material
+	BackgroundMaterialInstance = UMaterialInstanceDynamic::Create(BackgroundMaterial, nullptr);
 	SetHeight(Height);
 
-	MaterialInstance = UMaterialInstanceDynamic::Create(Material, nullptr);
-	Mesh->SetMaterial(0, MaterialInstance);
+	if (bInEditor)
+	{
+		// In editor, display the basic boundary material
+		Mesh->SetMaterial(0, BackgroundMaterialInstance);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/WorkerRegion.cpp:40]: Does this mean we're only seeing the text at runtime? Makes sense, although could be that seeing the virtual worker IDs would be useful, would be def out of scope for this PR anyway 👍

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 2]
```diff
@@ -33,9 +33,15 @@ DEFINE_LOG_CATEGORY(LogSpatialDebugger);
...
+// Background material for worker region
 const FString DEFAULT_WORKER_REGION_MATERIAL =
 	TEXT("/SpatialGDK/SpatialDebugger/Materials/TranslucentWorkerRegion.TranslucentWorkerRegion");
-}
+// Improbable primary font - Muli regular
+const FString DEFAULT_WORKER_TEXT_FONT = TEXT("/SpatialGDK/SpatialDebugger/Fonts/MuliFont.MuliFont");
+// Material to combine both the background and the worker information in one material
+const FString DEFAULT_WORKER_COMBINED_MATERIAL =
+	TEXT("/SpatialGDK/SpatialDebugger/Materials/WorkerRegionCombinedMaterial.WorkerRegionCombinedMaterial");
+} // namespace
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:44]:
```suggestion
} // anonymous namespace
```

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp [hunk 3]
```diff
@@ -164,19 +170,21 @@ void ASpatialDebugger::OnAuthorityGained()
...
 			for (int i = 0; i < LBStrategyRegions.Num(); i++)
 			{
-				const TPair<VirtualWorkerId, FBox2D>& LBStrategyRegion = LBStrategyRegions[i];
-				const PhysicalWorkerName* WorkerName =
-					NetDriver->VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(LBStrategyRegion.Key);
 				FWorkerRegionInfo WorkerRegionInfo;
-				WorkerRegionInfo.Color = (WorkerName == nullptr) ? InvalidServerTintColor : SpatialGDK::GetColorForWorkerName(*WorkerName);
+				const TPair<VirtualWorkerId, FBox2D>& LBStrategyRegion = LBStrategyRegions[i];
+				WorkerRegionInfo.VirtualWorkerID = LBStrategyRegion.Key;
+				WorkerRegionInfo.WorkerName = *NetDriver->VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(LBStrategyRegion.Key);
```
Comment (alastairdglennie) [inline SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp:176]: Previously we had a nullptr check, whereas here we're dereferencing without checking - are we confident there's no risk here?

**Review by alastairdglennie — APPROVED** [review]
lgtm, awesome

File: SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/WorkerRegion.cpp (hunk 1 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/WorkerRegion.cpp:40]: This approach to dynamically render the text on the material only works at runtime as the UCanvasRenderTarget2D->OnCanvasRenderTargetUpdate() callback is only triggered at runtime.

File: SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp (hunk 3 above)
Comment (Vix2021) [reply SpatialGDK/Source/SpatialGDK/Private/Utils/SpatialDebugger.cpp]: Good point. I will put the nullptr check back.

**Review by cmsmithio — APPROVED** [review]
Nice work!

**Comment by Vix2021** [comment]
Superceeded by https://github.com/spatialos/UnrealGDK/pull/2606

**Pull request closed without merging**